File: src/agent_search_1688/skills/catalog.py
"""Contained `SKILL.md` discovery and read-only access."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
import re


@dataclass(frozen=True)
class SkillEntry:
    name: str
    description: str
    root: Path


class SkillCatalog:
    def __init__(self, roots: list[Path]):
        self.roots = [root.resolve() for root in roots]

    def list(self) -> list[SkillEntry]:
        entries: dict[str, SkillEntry] = {}
        for root in self.roots:
            if not root.is_dir():
                continue
            for skill_file in root.glob("*/SKILL.md"):
                entry = self._entry(skill_file)
                if entry.name in entries:
                    raise ValueError(f"Skill 名称重复：{entry.name}")
                entries[entry.name] = entry
        return sorted(entries.values(), key=lambda entry: entry.name)

    def read(self, name: str, relative_path: str | None = None) -> str:
        entry = next((item for item in self.list() if item.name == name), None)
        if entry is None:
            raise KeyError(f"未找到项目 Skill：{name}")
        candidate = entry.root / (relative_path or "SKILL.md")
        resolved = candidate.resolve()
        if entry.root not in resolved.parents and resolved != entry.root:
            raise ValueError("Skill 引用文件超出 Skill 目录")
        if not resolved.is_file():
            raise FileNotFoundError("Skill 文件不存在")
        if resolved.stat().st_size > 200_000:
            raise ValueError("Skill 文件过大")
        return resolved.read_text(encoding="utf-8")

    @staticmethod
    def _entry(skill_file: Path) -> SkillEntry:
        text = skill_file.read_text(encoding="utf-8")
        if not text.startswith("---\n"):
            raise ValueError(f"Skill 缺少 YAML frontmatter：{skill_file}")
        frontmatter = text.split("---\n", 2)[1]
        values = dict(re.findall(r"^(name|description):\s*(.+)$", frontmatter, flags=re.MULTILINE))
        name = values.get("name", "").strip()
        description = values.get("description", "").strip()
        if not re.fullmatch(r"[a-z0-9][a-z0-9-]{0,63}", name) or not description:
            raise ValueError(f"Skill frontmatter 无效：{skill_file}")
        return SkillEntry(name=name, description=description, root=skill_file.parent.resolve())

File: src/skills/catalog.ts
import {
  lstatSync,
  readFileSync,
  readdirSync,
  realpathSync,
  statSync,
} from "node:fs";
import { isAbsolute, join, relative, resolve } from "node:path";

export interface SkillEntry {
  readonly name: string;
  readonly description: string;
  readonly root: string;
}

export class SkillCatalog {
  readonly #roots: readonly string[];

  constructor(roots: readonly string[]) {
    this.#roots = roots.map((root) => resolve(root));
  }

  list(): SkillEntry[] {
    const entries = new Map<string, SkillEntry>();
    for (const configuredRoot of this.#roots) {
      let root: string;
      try {
        if (!statSync(configuredRoot).isDirectory()) continue;
        root = realpathSync(configuredRoot);
      } catch {
        continue;
      }
      for (const directory of readdirSync(root, { withFileTypes: true })) {
        if (!directory.isDirectory()) continue;
        const skillFile = join(root, directory.name, "SKILL.md");
        try {
          if (!lstatSync(skillFile).isFile()) continue;
        } catch {
          continue;
        }
        const entry = parseSkillEntry(skillFile, root);
        if (entries.has(entry.name)) throw new Error(`Skill 名称重复：${entry.name}`);
        entries.set(entry.name, entry);
      }
    }
    return [...entries.values()].sort((left, right) => left.name.localeCompare(right.name));
  }

  read(name: string, relativePath = "SKILL.md"): string {
    const entry = this.list().find((item) => item.name === name);
    if (!entry) throw new Error(`未找到项目 Skill：${name}`);
    const candidate = resolve(entry.root, relativePath);
    assertContained(entry.root, candidate);
    let resolved: string;
    try {
      resolved = realpathSync(candidate);
    } catch (error) {
      throw new Error("Skill 文件不存在", { cause: error });
    }
    assertContained(entry.root, resolved);
    const stat = statSync(resolved);
    if (!stat.isFile()) throw new Error("Skill 文件不存在");
    if (stat.size > 200_000) throw new Error("Skill 文件过大");
    return readFileSync(resolved, "utf8");
  }
}

function parseSkillEntry(skillFile: string, catalogRoot: string): SkillEntry {
  const resolvedFile = realpathSync(skillFile);
  assertContained(catalogRoot, resolvedFile);
  const text = readFileSync(resolvedFile, "utf8");
  if (!text.startsWith("---\n")) throw new Error(`Skill 缺少 YAML frontmatter：${skillFile}`);
  const closing = text.indexOf("---\n", 4);
  if (closing < 0) throw new Error(`Skill 缺少 YAML frontmatter：${skillFile}`);
  const frontmatter = text.slice(4, closing);
  const name = frontmatter.match(/^name:\s*(.+)$/mu)?.[1]?.trim() ?? "";
  const description = frontmatter.match(/^description:\s*(.+)$/mu)?.[1]?.trim() ?? "";
  if (!/^[a-z0-9][a-z0-9-]{0,63}$/u.test(name) || !description) {
    throw new Error(`Skill frontmatter 无效：${skillFile}`);
  }
  return { name, description, root: realpathSync(resolve(resolvedFile, "..")) };
}

function assertContained(root: string, candidate: string): void {
  const child = relative(realpathSync(root), candidate);
  if (child === "" || (!isAbsolute(child) && child !== ".." && !child.startsWith(`..${process.platform === "win32" ? "\\" : "/"}`))) {
    return;
  }
  throw new Error("Skill 引用文件超出 Skill 目录");
}
